---
title: "10X analysis"
output: html_notebook
---

# Intro

## Data

Mail from Cédric.

> Les Long Reads :
>
> `/work/mzahm/Project_Pangafish/RawData/*.fastq.gz`
>
> Les reads Hi-C :
>
> `/work/mzahm/Project_Pangafish/RawData/HiC/*.fastq.gz`
>
> Les reads 10X :
>
> `/work/mzahm/Project_Pangafish/RawData/Run_PANGA1G.13800/Analyse_DemultiplexData.53847/*.fastq.gz`
>
> Le génome assemblé LR :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.fa`
>
> Le bam Longranger :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/ALIGN/outs/possorted_bam.bam`
>
> Le fichier de comptage des tags 10X par bin de 10k :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/possorted-tagcmp.tsv.gz`
>
> Le fichier “contact map” simulé :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.mnd.txt`
>
> Le fichier .hic généré à partir des données Hi-C :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.10X.hic`
>
> Le fichier .hic généré à partir des données 10X :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.Hi-C.hic`
>
> Le fichier .assembly du génome assemblé avec LR+Hi-C pour le chargement dans Juicebox :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.assembly`
>
>
> La procédure utilisée pour générer un fichier .hic à partir de données 10X :
>
> http://genoweb.toulouse.inra.fr/~sigenae/10X_and_Juicebox.html


 - Unpolished contigs: /work2/project/seqoccin/assemblies/assemblers/nanopore/bos_taurus/wtdbg2_trio1_mother_37165_run123/mother_raw.cgt.fa
 - ONT: /work2/project/seqoccin/data/reads/nanopore/bos_taurus/trio1.mother.run?.fastq.gz
 - Hi-C: /work2/project/seqoccin/assemblies/scaffolders/hic/bos_taurus/juicer_trio1_mother_37165_wtdbg2_ont_run123_hic_run1/Maison_plus/aligned/inter_30.hic
 - 10X: /work2/project/seqoccin/assemblies/scaffolders/10x/bos_taurus/LR_wtdbg_assembly_mother_37165_run123/Bovin-37165-align/outs/possorted_bam.bam
 - Assembly from JuiceBox: /work2/project/seqoccin/assemblies/scaffolders/hic/bos_taurus/juicer_trio1_mother_37165_wtdbg2_ont_run123_polished_hic_run1/Maison_plus/juicebox/mother_raw_wtdbg2_58x_polished.final.assembly

```{bash}
~/Desktop/Apps/minimap2/minimap2 -t 10 -o /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/GCF_002263795.1_ARS-UCD1.2_genomic.fna.gz /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa
~/Apps/samtools-1.10/samtools faidx /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa
cut -f 1-2 /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa.fai > /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len


~/Apps/samtools-1.10/samtools view -c /Scratch/mazytnicki/Mother_1/possorted_bam.bam
# 1590040440

../CheckScaffolding/checkBreaks /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len 1000000 100000 outBreaks.tsv
../CheckScaffolding/checkBreaks /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len breaks_hic.tsv 100000 10000 predicted.out putative.out > difference_breaks_hic.txt
../CheckScaffolding/checkBreaks /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len breaks_ont.tsv 100000 10000 predicted.out putative.out > difference_breaks_ont.txt
../CheckScaffolding/checkChromosome /Scratch/mazytnicki/Mother_1/mapping.paf > scaffChrs.txt
../CheckScaffolding/checkStitch scaffChrs.txt stitch.txt

./checkJoins /Scratch/mazytnicki/Mother_1/mapping.paf joins.txt 10000 10000

zcat /Scratch/mazytnicki/Mother_1/trio1.mother.run1.fastq.gz | wc -l
# 12064000

~/Desktop/Apps/minimap2/minimap2 -t 10 -o /Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa /Scratch/mazytnicki/Mother_1/trio1.mother.run1.fastq.gz

wc -l /Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf
# 4862048 /Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf
```

```{bash}
~/Apps/samtools-1.10/samtools index -@ 5 /Scratch/mazytnicki/Mother_1/possorted_bam.bam
~/Apps/samtools-1.10/samtools view -o /Scratch/mazytnicki/Mother_1/possorted_bam_reduced.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg10 ctg1061 ctg118 ctg1185 ctg12 ctg126 ctg128 ctg13 ctg1332 ctg136 ctg1427 ctg1435 ctg1451 ctg146 ctg154 ctg17 ctg173 ctg18 ctg189 ctg198 ctg2112 ctg219 ctg22 ctg23 ctg2379 ctg257 ctg26 ctg276 ctg281 ctg29 ctg30 ctg31 ctg312 ctg32 ctg322 ctg33 ctg3563 ctg3623 ctg38 ctg4 ctg4032 ctg4123 ctg4149 ctg43 ctg4330 ctg446 ctg4519 ctg4700 ctg4854 ctg4913 ctg50 ctg5008 ctg514 ctg516 ctg531 ctg549 ctg551 ctg5765 ctg6 ctg61 ctg655 ctg669 ctg67 ctg68 ctg703 ctg737 ctg766 ctg79 ctg8 ctg80 ctg82 ctg9 ctg91 ctg95 ctg976
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg4_ctg17.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg4 ctg17
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg2981_ctg1688.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg1688 ctg2981
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/small_test.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg104 ctg225
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg11.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg11
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg100.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg100
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg90.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg90
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg188.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg188
Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg37.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg37
```

```{r}
devtools::load_all(".")
resolution  <- 1000
bamFileName <- "/Scratch/mazytnicki/Mother_1/possorted_bam2.bam"
bamData     <- parseBamFile(bamFileName, resolution)
hicFileName <- "/Scratch/mazytnicki/Mother_1/inter_30.hic"
hicData     <- parseHicFile(hicFileName, resolution)
ontFileName <- "/Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf"
ontData     <- parsePafFile(ontFileName, resolution, 500, 10, 10)
contigFileName <- "/Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa"
scaffolder <- msscaf(contigFileName, resolution)
scaffolder <- addExp(scaffolder, ontData, "ONT")
scaffolder <- addExp(scaffolder, hicData, "HiC")
scaffolder <- addExp(scaffolder, bamData, "10X")
#saveRDS(scaffolder, file = "/Scratch/mazytnicki/Mother_1/allObject0.rds")
#scaffolder <- readRDS(file = "/Scratch/mazytnicki/Mother_1/allObject0.rds")
scaffolder <- estimateDistributions(scaffolder)
scaffolder <- cleanData(scaffolder)
#saveRDS(scaffolder, file = "/Scratch/mazytnicki/Mother_1/allObject.rds")
#scaffolder <- readRDS(file = "/Scratch/mazytnicki/Mother_1/allObject.rds")
scaffolder <- findBreaks(scaffolder, 0.01)
scaffolder <- splitChromosomes(scaffolder)
#saveRDS(scaffolder, file = "/Scratch/mazytnicki/Mother_1/allObject1.rds")
#scaffolder <- readRDS(file = "/Scratch/mazytnicki/Mother_1/allObject1.rds")
scaffolder <- findJoins(scaffolder, 0.05)
#saveRDS(scaffolder, file = "/Scratch/mazytnicki/Mother_1/allObject2.rds")
#scaffolder <- readRDS(file = "/Scratch/mazytnicki/Mother_1/allObject2.rds")
scaffolder <- scaffold(scaffolder)
Biostrings::writeXStringSet(Biostrings::DNAStringSet(scaffolder@sequences), "sequences.fa")
scaffolder@breaks %>% dplyr::select(ref, bin) %>% dplyr::mutate(bin = bin * resolution) %>% readr::write_tsv("breaks.tsv", col_names = FALSE)
scaffolder@joins %>% dplyr::arrange(pvalue) %>% dplyr::select(ref1, ref2) %>% readr::write_tsv("joins.tsv", col_names = FALSE)
```


```{bash}
../CheckScaffolding/verifyJoins /Scratch/mazytnicki/Mother_1/mapping.paf 100000 10000 joins.tsv
../CheckScaffolding/checkJoins /Scratch/mazytnicki/Mother_1/mapping.paf outJoins.tsv 1000 1000 joins.tsv
../CheckScaffolding/checkBreaks /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len 100000 10000 predicted.out breaks.tsv putative.out > difference_breaks.txt

# From tapou
~/Desktop/Apps/quast-5.0.2/quast-lg.py -t 15 -r /Scratch/mazytnicki/GCF_002263795.1_ARS-UCD1.2_genomic.fna.gz -t 6 -o /home/mazytnicki/Desktop/Projects/Other/Assembly/Quast/Test1 /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa wtdbg2_trio1_mother_37165_arcs_links.fa sequences.fa sequences_ONT_HiC.fa
# From genologin
```

```{bash}
scp /Scratch/mazytnicki/GCF_002263795.1_ARS-UCD1.2_genomic.fna.gz /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa wtdbg2_trio1_mother_37165_arcs_links.fa sequences.fa sequences_ONT_HiC.fa mzytnicki@genologin.toulouse.inrae.fr:/home/mzytnicki/work/Projects/tenxChecker
scp sequences.fa mzytnicki@genologin.toulouse.inrae.fr:/home/mzytnicki/work/Projects/tenxChecker
# From genologin
module load system/Miniconda3
module load bioinfo/Inspector-v1.2
cd /home/mzytnicki/work/Projects/tenxChecker
sbatch inspector.sh
```

```{r}
devtools::load_all(".")
#object <- makeSimObject(1, 1, 100, 20, 5, 0.2, c(1), c(50), c(), c(), c())
#object <- makeSimObject(2, 5, 100, 20, 10, 0.2, c(1, 2), c(25, 75), c(3, 4), c(4, 5), c(1, 5))
object <- makeSimObject(2, 5, 100, 100, 20, 0.5, c(1, 1, 2, 2), c(25, 75, 75, 75), c(3, 4), c(4, 5), c(1, 10))
#object <- makeSimObject(1, 2, 100, 20, 10, 0.2, c(), c(), c(1), c(2), c(5))
#plot.10X(object)
object <- estimateDistributions(object)
object <- cleanData(object)
object <- findBreaks(object, 0.05)
object <- splitChromosomes(object)
object <- findJoins(object, 0.05)
object <- scaffold(object)
```


```{r}
devtools::load_all(".")
allObject <- readRDS(file = "/Scratch/mazytnicki/Mother_1/allObject.rds")
splitObject <- generateSplit(allObject)
splitObject <- estimateDistributions(splitObject)
splitObject <- findJoins(splitObject, 0.05)
```


```{r}
devtools::load_all(".")
allObject <- readRDS(file = "/Scratch/mazytnicki/Mother_1/allObject.rds")
mergedObject <- generateJoin(allObject)
mergedObject <- estimateDistributions(mergedObject)
mergedObject <- findBreaks(mergedObject, 0.05)
```