---
title: "10X analysis"
output: html_notebook
---

# Intro

## Data

Mail from Cédric.

> Les Long Reads :
>
> `/work/mzahm/Project_Pangafish/RawData/*.fastq.gz`
>
> Les reads Hi-C :
>
> `/work/mzahm/Project_Pangafish/RawData/HiC/*.fastq.gz`
>
> Les reads 10X :
>
> `/work/mzahm/Project_Pangafish/RawData/Run_PANGA1G.13800/Analyse_DemultiplexData.53847/*.fastq.gz`
>
> Le génome assemblé LR :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.fa`
>
> Le bam Longranger :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/ALIGN/outs/possorted_bam.bam`
>
> Le fichier de comptage des tags 10X par bin de 10k :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/possorted-tagcmp.tsv.gz`
>
> Le fichier “contact map” simulé :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.mnd.txt`
>
> Le fichier .hic généré à partir des données Hi-C :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.10X.hic`
>
> Le fichier .hic généré à partir des données 10X :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.Hi-C.hic`
>
> Le fichier .assembly du génome assemblé avec LR+Hi-C pour le chargement dans Juicebox :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.assembly`
>
>
> La procédure utilisée pour générer un fichier .hic à partir de données 10X :
>
> http://genoweb.toulouse.inra.fr/~sigenae/10X_and_Juicebox.html


## Libraries

```{r}
library("MASS", pos = "package:base")
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
```


## Loading data

The complete dataset
```{r}
raw10X <- read_delim("/scratch/mazytnicki/possorted-tagcmp.tsv.gz",
                     delim     = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
# better save it...
saveRDS(object, "/Scratch/mazytnicki/bigData.rds")
object <- readRDS("/Scratch/mazytnicki/bigData.rds")
```

A small dataset, Scaff 001
```{r}
raw10X <- read_delim("possorted-tagcmp-short.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

Scaff 21
```{r}
raw10X <- read_delim("possorted-tagcmp-short-break2.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

Scaff 74
```{r}
raw10X <- read_delim("possorted-tagcmp-short-break3.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

Scaff 3
```{r}
raw10X <- read_delim("possorted-tagcmp-short-break4.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

A small dataset with scaffolding to do
```{r}
raw10X <- read_delim("possorted-tagcmp-short-orientation.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

Store into an object
```{r}
raw10X
object <- tenxcheckerExp(raw10X)
plot.10X(object)
```

## First steps

Clean data
```{r}
object <- estimateDistributions(object)
object <- cleanData(object)
gc()
```

```{r}
plot.10X(object)
```

Figures...
```{r}
p <- plotScaffoldSizeDistribution(object, log = FALSE)
p
```

```{r}
p <- plotCountDistribution(object, log = FALSE)
p
```

```{r}
p + xlim(-5, 100)
```

```{r}
plotCountDistribution(object, log = TRUE)
```

Plot intensity vs distance
```{r}
p <- plotMD(object, log = FALSE)
p
```

```{r}
p + xlim(0, 50) + ylim(0, 250)
    ```


```{r}
plotMD(object, log = TRUE)
```
```{r}
p <- plotRowCounts(object)
p
```

```{r}
plotRowCountDensity(object)
```

```{r}
p <- plotDiagonalStrength(object)
p
```


## Analysis

### Breaking chromosomes

```{r}
raw10X <- read_delim("possorted-tagcmp-short-break.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
object <- tenxcheckerExp(raw10X)
plot.10X(object)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
breaks <- checkBreaks(object)
filteredBreaks <- filterBreaks(object, breaks$data)
object <- splitChromosomes(object, filteredBreaks$breaks)
plot.10X(object)
```

### Scaffolding chromosomes

```{r}
raw10X <- read_delim("possorted-tagcmp-short-orientation.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
object <- tenxcheckerExp(raw10X)
plot.10X(object)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
joins <- checkJoins(object)
joins <- filterJoins(object, joins)
object <- scaffold(object, joins)
plot.10X(object)
```

### Finding short insertions

```{r}
insertions <- checkInsertions(object)
insertions$data %>% 
    filter(triangle > 0.5) %>%
    filter(rest < -0.5)
```
   ref             bin distance triangle  rest difference
   <fct>         <int>    <int>    <dbl> <dbl>      <dbl>
 1 PAHY_Scaf_069    80        2    0.825 -4.50       5.33
 2 PAHY_Scaf_069    84        2    0.898 -4.43       5.33
 3 PAHY_Scaf_069    79        3    0.610 -4.47       5.08
 4 PAHY_Scaf_069    85        3    0.693 -4.48       5.17
 5 PAHY_Scaf_069    78        4    0.507 -4.48       4.99
 6 PAHY_Scaf_069    86        4    0.553 -4.44       4.99
 7 PAHY_Scaf_088     5        2    1.01  -1.20       2.21
 8 PAHY_Scaf_088     6        3    0.633 -1.60       2.24
 9 PAHY_Scaf_101     5        2    0.765 -4.50       5.26
10 PAHY_Scaf_101     6        3    0.596 -4.52       5.11
11 PAHY_Scaf_154    41        2    0.634 -4.57       5.20
12 PAHY_Scaf_154    40        3    0.527 -2.50       3.03
13 PAHY_Scaf_174     7        4    0.536 -5.15       5.68

```{r}
raw10X <- read_delim("possorted-tagcmp-088.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
object <- tenxcheckerExp(raw10X)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
plotInsertions2(extractRef(object, "PAHY_Scaf_088"), 5, 2)
```


```{r}
raw10X <- read_delim("possorted-tagcmp-069.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
object <- tenxcheckerExp(raw10X)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
plotInsertions2(extractRef(object, "PAHY_Scaf_069"), 78, 4)
```

```{r}
raw10X <- parseBamFile("phased_possorted_bam.bam", 40000)
object <- tenxcheckerExp(raw10X)
plot.10X(object)
```

```{r}
#raw10X <- parseBamFile("/home/mazytnicki/Genotoul/Panga_longranger/map_to_scaffolds/ALIGN/outs/possorted_bam.bam", 1000000)
#raw10X <- parseBamFile("/home/mazytnicki/Genotoul/seqoccin/workspace/10X_Matthias/20kb/ALIGN/outs/TestFilter.bam", 100000)
raw10X <- parseBamFile("PAHY_Scaf_001.bam", 1000)
raw10X <- parseBamFile("PAHY_Scaf_104.bam", 1000)
object <- tenxcheckerExp(raw10X, 10000)
plot.10X(object)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
breaks <- checkBreaks(object)
filteredBreaks <- filterBreaks(object, breaks$data)
object <- splitChromosomes(object, filteredBreaks$breaks)
plot.10X(object)

joins <- checkJoins(object)
joins <- filterJoins(object, joins)
object <- scaffold(object, joins)
plot.10X(object)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
hicFileName <- "/home/mazytnicki/Genotoul/seqoccin/assemblies/assemblers/mixed/bos_taurus/Trio2/test/hic/inter_30.hic"
hicFileName <- "/Scratch/mazytnicki/inter_30.hic"
#hicFileName <- "/home/mazytnicki/Genotoul/seqoccin/assemblies/assemblers/mixed/bos_taurus/Trio2/test/hic/3ddna.rawchrom.hic"
#hicData <- parseHicFile(hicFileName, 50000)
hicData <- parseHicFile(hicFileName, 10000)
saveRDS(hicData, "/media/Stock/tenxchecker/hic.rds")
hicData <- readRDS("/media/Stock/tenxchecker/hic.rds")
hicObject <- tenxcheckerExp(hicData)
hicObject <- estimateDistributions(hicObject)
hicObject@parameters@sampleSize   <- 100000
hicObject@parameters@minCount     <- 3
hicObject@parameters@maxLinkRange <- 500
hicObject@parameters@minRowCount  <- 10
hicObject@parameters@minNBins     <- 2
hicObject@parameters@loessSpan    <- 0.5
hicObject <- cleanData(hicObject)
objects <- splitByRef(hicObject)
hicBreaks <- checkBreaks(hicObject)
hicObject@parameters@breakNCells    <- 100
hicObject@parameters@breakThreshold <- -0.6
filteredHicBreaks <- filterBreaks(hicObject, hicBreaks$data)
hicObject <- splitChromosomes(hicObject, filteredHicBreaks$breaks)
hicObject@parameters@maxLinkRange <- 10
hicObject@parameters@breakNCells <- 10
joins <- checkJoins(hicObject)
filteredJoins <- filterJoins(hicObject, joins)
hicObject <- scaffold(hicObject, joins)
plot.10X(hicObject)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
register(MulticoreParam(workers = 10), default = TRUE)
devtools::load_all(".")
bamFileName <- "/Scratch/mazytnicki/bovin-37163.bam"
#bamFileName <- "/media/Stock/bovin-37163.bam"
#bamFileName <- "/media/Stock/bovin-37163-S1.bam"
#bamFileName <- "/media/Stock/test.bam"
bamFileName <- "/media/Stock/test.bam"
bamFileName <- "/media/Stock/smallTest.bam"
bamFileName <- "/media/Stock/verySmallTest.bam"
bamFileName       <- "/Scratch/mazytnicki/test.bam"
#bamFileName       <- "/Scratch/mazytnicki/smallTest.bam"
#bamFileName       <- "/Scratch/mazytnicki/verySmallTest.bam"
bamData           <- parseBamFile(bamFileName, 10000, 12)
bamObject         <- tenxcheckerExp(bamData)
#plotBackgroundCountDistribution(bamObject)
#plotMoleculeSizeDistribution(bamObject)
#plotRowCountDistribution(bamObject)
bamObject         <- estimateDistributions(bamObject)
bamObject@parameters@minCount <- 5
bamObject@parameters@maxLinkRange <- 10
bamObject@parameters@minRowCount <- 100
bamObject@parameters@minNBins <- 2
bamObject@parameters@maxLinkRange   <- 10
bamObject@parameters@breakThreshold <- -3
bamObject@parameters@breakNCells <- 0.75 * ((bamObject@parameters@maxLinkRange * (bamObject@parameters@maxLinkRange + 1)) / 2)
bamObject         <- cleanData(bamObject)
bamBreaks         <- checkBreaks(bamObject)
filteredBamBreaks <- filterBreaks(bamObject, bamBreaks$data)
bamObject <- splitChromosomes(bamObject, filteredBamBreaks$breaks)
joins <- checkJoins(bamObject)
joins <- filterJoins(bamObject, joins)
bamObject <- scaffold(bamObject, joins)
plot.10X(object)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
hicFileName <- "/media/Stock/genome.final.10X.hic"
hicData <- parseHicFile(hicFileName, 10000)
hicObject <- tenxcheckerExp(hicData, 10000)
saveRDS(ontData, "/media/Stock/tenxchecker/10x.rds")
hicObject <- estimateDistributions(hicObject)
hicObject <- cleanData(hicObject)
hicObjects <- splitByRef(hicObject)
plot.10X(hicObjects)
hicBreaks <- checkBreaks(hicObject)
filteredHicBreaks <- filterBreaks(hicObject, hicBreaks$data)
hicObject <- splitChromosomes(hicObject, filteredHicBreaks$breaks)
joins <- checkJoins(hicObject)
joins <- filterJoins(hicObject, joins)
hicObject <- scaffold(hicObject, joins)
plot.10X(object)
```

```{r}
library("MASS", pos = "package:base")
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
library("tibble")
devtools::load_all(".")
ontFileName <- "/home/mazytnicki/Genotoul/seqoccin/assemblies/assemblers/mixed/bos_taurus/Trio2/test/nanopore/aln12345.paf"
ontFileName <- "/Scratch/mazytnicki/aln12345.paf"
ontData <- parsePafFile(ontFileName, 1000, 500, 10, 10)
ontObject <- tenxcheckerExp(ontData)
ontObject <- estimateDistributions(ontObject)
# ontObject@parameters@minCount <- 2
# ontObject@parameters@minRowCount <- 2
# ontObject@parameters@minNBins <- 2
plotBackgroundCountDistribution(ontObject)
plotMoleculeSizeDistribution(ontObject)
plotRowCountDistribution(ontObject)
ontObject <- cleanData(ontObject)
objects <- splitByRef(ontObject)
ontBreaks <- checkBreaks(ontObject)
filteredOntBreaks <- filterBreaks(ontObject, ontBreaks$data)
ontObject <- splitChromosomes(ontObject, filteredOntBreaks$breaks)
joins <- checkJoins(ontObject)
joins <- filterJoins(ontObject, joins)
ontObject <- scaffold(ontObject, joins)
```

```{r}
library("MASS", pos = "package:base")
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
referenceFileName <- "/home/mazytnicki/Genotoul/seqoccin/assemblies/assemblers/mixed/bos_taurus/Trio2/test/reference/alnContigsReference.paf"
referenceData <- parsePafFile(referenceFileName, 100000, 1000, 1, 1)
saveRDS(referenceData, "/media/Stock/tenxchecker/reference.rds")
referenceData
```

```{r}
object <- ontObject
object <- bamObject
object <- hicObject
objects <- splitByRef(object)
plot.10XRef(objects[[12]], lim = c(0, 100))
l <- estimateBackgroundCounts(object)
object@parameters@minCount <- l$count
l <- estimateMoleculeSize(object)
object@parameters@maxLinkRange <- l$size
object@parameters@breakNCells <- 0.75 * ((object@parameters@maxLinkRange * (object@parameters@maxLinkRange + 1)) / 2)
l <- estimateMinRowCount(object)
object@parameters@minRowCount <- l$count
```

 - Unpolished contigs: /work2/project/seqoccin/assemblies/assemblers/nanopore/bos_taurus/wtdbg2_trio1_mother_37165_run123/mother_raw.cgt.fa
 - ONT: /work2/project/seqoccin/data/reads/nanopore/bos_taurus/trio1.mother.run?.fastq.gz
 - Hi-C: /work2/project/seqoccin/assemblies/scaffolders/hic/bos_taurus/juicer_trio1_mother_37165_wtdbg2_ont_run123_hic_run1/Maison_plus/aligned/inter_30.hic
 - 10X: /work2/project/seqoccin/assemblies/scaffolders/10x/bos_taurus/LR_wtdbg_assembly_mother_37165_run123/Bovin-37165-align/outs/possorted_bam.bam

```{bash}
~/Desktop/Apps/minimap2/minimap2 -t 10 -o /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/GCF_002263795.1_ARS-UCD1.2_genomic.fna.gz /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa
~/Apps/samtools-1.10/samtools faidx /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa
cut -f 1-2 /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa.fai > /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len


~/Apps/samtools-1.10/samtools view -c /Scratch/mazytnicki/Mother_1/possorted_bam.bam
# 1590040440

./checkBreaks /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len breaks_hic.tsv 100000 10000 predicted.out putative.out > d

./checkJoins /Scratch/mazytnicki/Mother_1/mapping.paf joins.txt 10000 10000

zcat /Scratch/mazytnicki/Mother_1/trio1.mother.run1.fastq.gz | wc -l
# 12064000

~/Desktop/Apps/minimap2/minimap2 -t 10 -o /Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa /Scratch/mazytnicki/Mother_1/trio1.mother.run1.fastq.gz

wc -l /Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf
# 4862048 /Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
hicFileName <- "/Scratch/mazytnicki/Mother_1/inter_30.hic"
hicData <- parseHicFile(hicFileName, 10000)
hicObject <- tenxcheckerExp(hicData)
hicObject <- estimateDistributions(hicObject)
hicObject@parameters@sampleSize   <- 100000
hicObject@parameters@minCount     <- 3
hicObject@parameters@maxLinkRange <- 500
hicObject@parameters@minRowCount  <- 10
hicObject@parameters@minNBins     <- 2
hicObject@parameters@loessSpan    <- 0.5
hicObject <- cleanData(hicObject)
hicBreaks <- checkBreaks(hicObject)
hicObject@parameters@breakNCells    <- 30
hicObject@parameters@breakThreshold <- -0.3
filteredHicBreaks <- filterBreaks(hicObject, hicBreaks$data)
filteredHicBreaks$breaks %>% mutate(pos = bin * hicObject@parameters@binSize) %>% mutate(ref = str_to_lower(ref)) %>% select(ref, pos) %>% write_tsv("breaks_hic.tsv", col_names = FALSE)
hicObject <- splitChromosomes(hicObject, filteredHicBreaks$breaks)
hicObject@parameters@maxLinkRange <- 10
hicObject@parameters@breakNCells <- 10
joins <- checkJoins(hicObject)
filteredJoins <- filterJoins(hicObject, joins)
hicObject <- scaffold(hicObject, filteredJoins)
plot.10X(hicObject)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
register(MulticoreParam(workers = 10), default = TRUE)
devtools::load_all(".")
bamFileName <- "/Scratch/mazytnicki/Mother_1/possorted_bam.bam"
bamData           <- parseBamFile(bamFileName, 10000, 12)
bamObject         <- tenxcheckerExp(bamData)
#plotBackgroundCountDistribution(bamObject)
#plotMoleculeSizeDistribution(bamObject)
#plotRowCountDistribution(bamObject)
bamObject         <- estimateDistributions(bamObject)
bamObject@parameters@minCount <- 5
bamObject@parameters@maxLinkRange <- 10
bamObject@parameters@minRowCount <- 100
bamObject@parameters@minNBins <- 2
bamObject@parameters@maxLinkRange   <- 10
bamObject@parameters@breakThreshold <- -3
bamObject@parameters@breakNCells <- 0.75 * ((bamObject@parameters@maxLinkRange * (bamObject@parameters@maxLinkRange + 1)) / 2)
bamObject         <- cleanData(bamObject)
bamBreaks         <- checkBreaks(bamObject)
filteredBamBreaks <- filterBreaks(bamObject, bamBreaks$data)
bamObject <- splitChromosomes(bamObject, filteredBamBreaks$breaks)
joins <- checkJoins(bamObject)
joins <- filterJoins(bamObject, joins)
bamObject <- scaffold(bamObject, joins)
plot.10X(object)
```

```{r}
library("MASS", pos = "package:base")
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
library("tibble")
devtools::load_all(".")
ontFileName <- "/Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf"
ontData <- parsePafFile(ontFileName, 10000, 500, 10, 10)
ontObject <- tenxcheckerExp(ontData)
ontObject <- estimateDistributions(ontObject)
ontObject@parameters@minCount <- 2
ontObject@parameters@minRowCount <- 2
ontObject@parameters@minNBins <- 2
plotBackgroundCountDistribution(ontObject)
plotMoleculeSizeDistribution(ontObject)
plotRowCountDistribution(ontObject)
ontObject <- cleanData(ontObject)
ontBreaks <- checkBreaks(ontObject, FALSE, FALSE, TRUE)
ontObject@parameters@maxLinkRange <- 2
filteredOntBreaks <- filterBreaks(ontObject, ontBreaks$data)
filteredHicBreaks$breaks %>% mutate(pos = bin * hicObject@parameters@binSize) %>% mutate(ref = str_to_lower(ref)) %>% select(ref, pos) %>% write_tsv("breaks_hic.tsv", col_names = FALSE)
ontObject <- splitChromosomes(ontObject, filteredOntBreaks$breaks)
joins <- checkJoins(ontObject)
joins <- filterJoins(ontObject, joins)
ontObject <- scaffold(ontObject, joins)
```
