---
title: "10X analysis"
output: html_notebook
---

# Intro

## Data

Mail from Cédric.

> Les Long Reads :
>
> `/work/mzahm/Project_Pangafish/RawData/*.fastq.gz`
>
> Les reads Hi-C :
>
> `/work/mzahm/Project_Pangafish/RawData/HiC/*.fastq.gz`
>
> Les reads 10X :
>
> `/work/mzahm/Project_Pangafish/RawData/Run_PANGA1G.13800/Analyse_DemultiplexData.53847/*.fastq.gz`
>
> Le génome assemblé LR :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.fa`
>
> Le bam Longranger :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/ALIGN/outs/possorted_bam.bam`
>
> Le fichier de comptage des tags 10X par bin de 10k :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/possorted-tagcmp.tsv.gz`
>
> Le fichier “contact map” simulé :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.mnd.txt`
>
> Le fichier .hic généré à partir des données Hi-C :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.10X.hic`
>
> Le fichier .hic généré à partir des données 10X :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.Hi-C.hic`
>
> Le fichier .assembly du génome assemblé avec LR+Hi-C pour le chargement dans Juicebox :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.assembly`
>
>
> La procédure utilisée pour générer un fichier .hic à partir de données 10X :
>
> http://genoweb.toulouse.inra.fr/~sigenae/10X_and_Juicebox.html


## Libraries

```{r}
library("MASS", pos = "package:base")
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
```


## Loading data

The complete dataset
```{r}
# raw10X <- read_delim("/scratch/mazytnicki/possorted-tagcmp.tsv.gz",
#                      delim = " ",
#                      col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
#                      col_types = "fifii")
# #better save it...
# saveRDS(object, "/Scratch/mazytnicki/bigData.rds")
object <- readRDS("/Scratch/mazytnicki/bigData.rds")
```

A small dataset
```{r}
raw10X <- read_delim("possorted-tagcmp-short.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

A small dataset with a break
```{r}
raw10X <- read_delim("possorted-tagcmp-short-break.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

```{r}
raw10X <- read_delim("possorted-tagcmp-short-break2.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

```{r}
raw10X <- read_delim("possorted-tagcmp-short-break3.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

A small dataset with a reversal
```{r}
raw10X <- read_delim("possorted-tagcmp-short-orientation.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

Store into an object
```{r}
raw10X
object <- tenxcheckerExp(raw10X)
plot.10X(object)
```

## First steps

Clean data
```{r}
object <- estimateDistributions(object)
object <- cleanData(object)
```

```{r}
plot.10X(object)
```

Figures...
```{r}
p <- plotScaffoldSizeDistribution(object, log = FALSE)
p
```

```{r}
p <- plotCountDistribution(object, log = FALSE)
p
```

```{r}
p + xlim(-5, 100)
```

```{r}
plotCountDistribution(object, log = TRUE)
```

Plot intensity vs distance
```{r}
p <- plotMD(object, log = FALSE)
p
```

```{r}
p + xlim(0, 50) + ylim(0, 250)
    ```


```{r}
plotMD(object, log = TRUE)
```
```{r}
p <- plotRowCounts(object)
p
```

```{r}
plotRowCountDensity(object)
```

```{r}
p <- plotDiagonalStrength(object)
```


## Analysis

```{r}
breaks <- checkBreaks(object)
```

```{r}
joins <- checkJoins(object)
j <- joins %>%
    mutate(pvalue = map_dbl(test, "p.value")) %>%
    separate(key, c(1, 2), into = c(NA, "vert", "hor")) %>%
    mutate(vert = factor(vert)) %>%
    mutate(hor = factor(hor))
j <- j %>% select(ref1, ref2, vert, hor, pvalue)
j1 <- j %>%
    filter(hor == "L") %>%
    rename(ref = ref1, otherRef = ref2) %>%
    mutate(place = "before") %>%
    select(-c(vert, hor))
j2 <- j %>%
    filter(hor == "R") %>%
    rename(ref = ref1, otherRef = ref2) %>%
    mutate(place = "after") %>%
    select(-c(vert, hor))
j3 <- j %>%
    filter(vert == "B") %>%
    rename(ref = ref2, otherRef = ref1) %>%
    mutate(place = "after") %>%
    select(-c(vert, hor))
j4 <- j %>%
    filter(vert == "U") %>%
    rename(ref = ref2, otherRef = ref1) %>%
    mutate(place = "before") %>%
    select(-c(vert, hor))
js <- bind_rows(j1, j2, j3, j4) %>% mutate(place = factor(place))
jst <- js %>%
    group_by(ref, place) %>%
    filter(pvalue == min(pvalue)) %>%
    ungroup()
jsts <- bind_rows(jst %>% filter(place == "before") %>%
                      rename(firstRef = ref, secondRef = otherRef),
                  jst %>% filter(place == "after") %>% 
                      rename(firstRef = otherRef, secondRef = ref)) %>%
    select(-place) %>%
    group_by(firstRef) %>% filter(pvalue == min(pvalue)) %>% ungroup() %>%
    group_by(secondRef) %>% filter(pvalue == min(pvalue)) %>% ungroup()
```

