---
title: "10X analysis"
output: html_notebook
---

# Intro

## Data

Mail from Cédric.

> Les Long Reads :
>
> `/work/mzahm/Project_Pangafish/RawData/*.fastq.gz`
>
> Les reads Hi-C :
>
> `/work/mzahm/Project_Pangafish/RawData/HiC/*.fastq.gz`
>
> Les reads 10X :
>
> `/work/mzahm/Project_Pangafish/RawData/Run_PANGA1G.13800/Analyse_DemultiplexData.53847/*.fastq.gz`
>
> Le génome assemblé LR :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.fa`
>
> Le bam Longranger :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/ALIGN/outs/possorted_bam.bam`
>
> Le fichier de comptage des tags 10X par bin de 10k :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/possorted-tagcmp.tsv.gz`
>
> Le fichier “contact map” simulé :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.mnd.txt`
>
> Le fichier .hic généré à partir des données Hi-C :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.10X.hic`
>
> Le fichier .hic généré à partir des données 10X :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.Hi-C.hic`
>
> Le fichier .assembly du génome assemblé avec LR+Hi-C pour le chargement dans Juicebox :
>
> `/work/project/sigenae/Christophe/Panga_longranger/map_to_scaffolds/genome.final.assembly`
>
>
> La procédure utilisée pour générer un fichier .hic à partir de données 10X :
>
> http://genoweb.toulouse.inra.fr/~sigenae/10X_and_Juicebox.html


## Libraries

```{r}
library("MASS", pos = "package:base")
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
```


## Loading data

The complete dataset
```{r}
raw10X <- read_delim("/scratch/mazytnicki/possorted-tagcmp.tsv.gz",
                     delim     = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
# better save it...
saveRDS(object, "/Scratch/mazytnicki/bigData.rds")
object <- readRDS("/Scratch/mazytnicki/bigData.rds")
```

A small dataset, Scaff 001
```{r}
raw10X <- read_delim("possorted-tagcmp-short.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

Scaff 21
```{r}
raw10X <- read_delim("possorted-tagcmp-short-break2.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

Scaff 74
```{r}
raw10X <- read_delim("possorted-tagcmp-short-break3.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

Scaff 3
```{r}
raw10X <- read_delim("possorted-tagcmp-short-break4.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

A small dataset with scaffolding to do
```{r}
raw10X <- read_delim("possorted-tagcmp-short-orientation.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
```

Store into an object
```{r}
raw10X
object <- tenxcheckerExp(raw10X)
plot.10X(object)
```

## First steps

Clean data
```{r}
object <- estimateDistributions(object)
object <- cleanData(object)
gc()
```

```{r}
plot.10X(object)
```

Figures...
```{r}
p <- plotScaffoldSizeDistribution(object, log = FALSE)
p
```

```{r}
p <- plotCountDistribution(object, log = FALSE)
p
```

```{r}
p + xlim(-5, 100)
```

```{r}
plotCountDistribution(object, log = TRUE)
```

Plot intensity vs distance
```{r}
p <- plotMD(object, log = FALSE)
p
```

```{r}
p + xlim(0, 50) + ylim(0, 250)
    ```


```{r}
plotMD(object, log = TRUE)
```
```{r}
p <- plotRowCounts(object)
p
```

```{r}
plotRowCountDensity(object)
```

```{r}
p <- plotDiagonalStrength(object)
p
```


## Analysis

### Breaking chromosomes

```{r}
raw10X <- read_delim("possorted-tagcmp-short-break.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
object <- tenxcheckerExp(raw10X)
plot.10X(object)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
breaks <- checkBreaks(object)
filteredBreaks <- filterBreaks(object, breaks$data)
object <- splitChromosomes(object, filteredBreaks$breaks)
plot.10X(object)
```

### Scaffolding chromosomes

```{r}
raw10X <- read_delim("possorted-tagcmp-short-orientation.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
object <- tenxcheckerExp(raw10X)
plot.10X(object)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
joins <- checkJoins(object)
joins <- filterJoins(object, joins)
object <- scaffold(object, joins)
plot.10X(object)
```

### Finding short insertions

```{r}
insertions <- checkInsertions(object)
insertions$data %>% 
    filter(triangle > 0.5) %>%
    filter(rest < -0.5)
```
   ref             bin distance triangle  rest difference
   <fct>         <int>    <int>    <dbl> <dbl>      <dbl>
 1 PAHY_Scaf_069    80        2    0.825 -4.50       5.33
 2 PAHY_Scaf_069    84        2    0.898 -4.43       5.33
 3 PAHY_Scaf_069    79        3    0.610 -4.47       5.08
 4 PAHY_Scaf_069    85        3    0.693 -4.48       5.17
 5 PAHY_Scaf_069    78        4    0.507 -4.48       4.99
 6 PAHY_Scaf_069    86        4    0.553 -4.44       4.99
 7 PAHY_Scaf_088     5        2    1.01  -1.20       2.21
 8 PAHY_Scaf_088     6        3    0.633 -1.60       2.24
 9 PAHY_Scaf_101     5        2    0.765 -4.50       5.26
10 PAHY_Scaf_101     6        3    0.596 -4.52       5.11
11 PAHY_Scaf_154    41        2    0.634 -4.57       5.20
12 PAHY_Scaf_154    40        3    0.527 -2.50       3.03
13 PAHY_Scaf_174     7        4    0.536 -5.15       5.68

```{r}
raw10X <- read_delim("possorted-tagcmp-088.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
object <- tenxcheckerExp(raw10X)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
plotInsertions2(extractRef(object, "PAHY_Scaf_088"), 5, 2)
```


```{r}
raw10X <- read_delim("possorted-tagcmp-069.tsv",
                     delim = " ",
                     col_names = c("ref1", "bin1", "ref2", "bin2", "count"),
                     col_types = "fifii")
object <- tenxcheckerExp(raw10X)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
plotInsertions2(extractRef(object, "PAHY_Scaf_069"), 78, 4)
```

```{r}
raw10X <- parseBamFile("phased_possorted_bam.bam", 40000)
object <- tenxcheckerExp(raw10X)
plot.10X(object)
```

```{r}
#raw10X <- parseBamFile("/home/mazytnicki/Genotoul/Panga_longranger/map_to_scaffolds/ALIGN/outs/possorted_bam.bam", 1000000)
#raw10X <- parseBamFile("/home/mazytnicki/Genotoul/seqoccin/workspace/10X_Matthias/20kb/ALIGN/outs/TestFilter.bam", 100000)
raw10X <- parseBamFile("PAHY_Scaf_001.bam", 1000)
raw10X <- parseBamFile("PAHY_Scaf_104.bam", 1000)
object <- tenxcheckerExp(raw10X, 10000)
plot.10X(object)
object <- estimateDistributions(object)
object <- cleanData(object)
plot.10X(object)
breaks <- checkBreaks(object)
filteredBreaks <- filterBreaks(object, breaks$data)
object <- splitChromosomes(object, filteredBreaks$breaks)
plot.10X(object)

joins <- checkJoins(object)
joins <- filterJoins(object, joins)
object <- scaffold(object, joins)
plot.10X(object)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
hicFileName <- "/home/mazytnicki/Genotoul/seqoccin/assemblies/assemblers/mixed/bos_taurus/Trio2/test/hic/inter_30.hic"
hicFileName <- "/Scratch/mazytnicki/inter_30.hic"
#hicFileName <- "/home/mazytnicki/Genotoul/seqoccin/assemblies/assemblers/mixed/bos_taurus/Trio2/test/hic/3ddna.rawchrom.hic"
#hicData <- parseHicFile(hicFileName, 50000)
hicData <- parseHicFile(hicFileName, 10000)
saveRDS(hicData, "/media/Stock/tenxchecker/hic.rds")
hicData <- readRDS("/media/Stock/tenxchecker/hic.rds")
hicObject <- tenxcheckerExp(hicData)
hicObject <- estimateDistributions(hicObject)
hicObject@parameters@sampleSize   <- 100000
hicObject@parameters@minCount     <- 3
hicObject@parameters@maxLinkRange <- 500
hicObject@parameters@minRowCount  <- 10
hicObject@parameters@minNBins     <- 2
hicObject@parameters@loessSpan    <- 0.5
hicObject <- cleanData(hicObject)
objects <- splitByRef(hicObject)
hicBreaks <- checkBreaks(hicObject)
hicObject@parameters@breakNCells    <- 100
hicObject@parameters@breakThreshold <- -0.6
filteredHicBreaks <- filterBreaks(hicObject, hicBreaks$data)
hicObject <- splitChromosomes(hicObject, filteredHicBreaks$breaks)
hicObject@parameters@maxLinkRange <- 10
hicObject@parameters@breakNCells <- 10
joins <- checkJoins(hicObject)
filteredJoins <- filterJoins(hicObject, joins)
hicObject <- scaffold(hicObject, joins)
plot.10X(hicObject)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
register(MulticoreParam(workers = 10), default = TRUE)
devtools::load_all(".")
bamFileName <- "/Scratch/mazytnicki/bovin-37163.bam"
#bamFileName <- "/media/Stock/bovin-37163.bam"
#bamFileName <- "/media/Stock/bovin-37163-S1.bam"
#bamFileName <- "/media/Stock/test.bam"
bamFileName <- "/media/Stock/test.bam"
bamFileName <- "/media/Stock/smallTest.bam"
bamFileName <- "/media/Stock/verySmallTest.bam"
bamFileName       <- "/Scratch/mazytnicki/test.bam"
#bamFileName       <- "/Scratch/mazytnicki/smallTest.bam"
#bamFileName       <- "/Scratch/mazytnicki/verySmallTest.bam"
bamData           <- parseBamFile(bamFileName, 10000, 12)
bamObject         <- tenxcheckerExp(bamData)
#plotBackgroundCountDistribution(bamObject)
#plotMoleculeSizeDistribution(bamObject)
#plotRowCountDistribution(bamObject)
bamObject         <- estimateDistributions(bamObject)
bamObject@parameters@minCount <- 5
bamObject@parameters@maxLinkRange <- 10
bamObject@parameters@minRowCount <- 100
bamObject@parameters@minNBins <- 2
bamObject@parameters@maxLinkRange   <- 10
bamObject@parameters@breakThreshold <- -3
bamObject@parameters@breakNCells <- 0.75 * ((bamObject@parameters@maxLinkRange * (bamObject@parameters@maxLinkRange + 1)) / 2)
bamObject         <- cleanData(bamObject)
bamBreaks         <- checkBreaks(bamObject)
filteredBamBreaks <- filterBreaks(bamObject, bamBreaks$data)
bamObject <- splitChromosomes(bamObject, filteredBamBreaks$breaks)
joins <- checkJoins(bamObject)
joins <- filterJoins(bamObject, joins)
bamObject <- scaffold(bamObject, joins)
plot.10X(object)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
hicFileName <- "/media/Stock/genome.final.10X.hic"
hicData <- parseHicFile(hicFileName, 10000)
hicObject <- tenxcheckerExp(hicData, 10000)
saveRDS(ontData, "/media/Stock/tenxchecker/10x.rds")
hicObject <- estimateDistributions(hicObject)
hicObject <- cleanData(hicObject)
hicObjects <- splitByRef(hicObject)
plot.10X(hicObjects)
hicBreaks <- checkBreaks(hicObject)
filteredHicBreaks <- filterBreaks(hicObject, hicBreaks$data)
hicObject <- splitChromosomes(hicObject, filteredHicBreaks$breaks)
joins <- checkJoins(hicObject)
joins <- filterJoins(hicObject, joins)
hicObject <- scaffold(hicObject, joins)
plot.10X(object)
```

```{r}
library("MASS", pos = "package:base")
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
library("tibble")
devtools::load_all(".")
ontFileName <- "/home/mazytnicki/Genotoul/seqoccin/assemblies/assemblers/mixed/bos_taurus/Trio2/test/nanopore/aln12345.paf"
ontFileName <- "/Scratch/mazytnicki/aln12345.paf"
ontData <- parsePafFile(ontFileName, 1000, 500, 10, 10)
ontObject <- tenxcheckerExp(ontData)
ontObject <- estimateDistributions(ontObject)
# ontObject@parameters@minCount <- 2
# ontObject@parameters@minRowCount <- 2
# ontObject@parameters@minNBins <- 2
plotBackgroundCountDistribution(ontObject)
plotMoleculeSizeDistribution(ontObject)
plotRowCountDistribution(ontObject)
ontObject <- cleanData(ontObject)
objects <- splitByRef(ontObject)
ontBreaks <- checkBreaks(ontObject)
filteredOntBreaks <- filterBreaks(ontObject, ontBreaks$data)
ontObject <- splitChromosomes(ontObject, filteredOntBreaks$breaks)
joins <- checkJoins(ontObject)
joins <- filterJoins(ontObject, joins)
ontObject <- scaffold(ontObject, joins)
```

```{r}
library("MASS", pos = "package:base")
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
referenceFileName <- "/home/mazytnicki/Genotoul/seqoccin/assemblies/assemblers/mixed/bos_taurus/Trio2/test/reference/alnContigsReference.paf"
referenceData <- parsePafFile(referenceFileName, 100000, 1000, 1, 1)
saveRDS(referenceData, "/media/Stock/tenxchecker/reference.rds")
referenceData
```

```{r}
object <- ontObject
object <- bamObject
object <- hicObject
objects <- splitByRef(object)
plot.10XRef(objects[[12]], lim = c(0, 100))
l <- estimateBackgroundCounts(object)
object@parameters@minCount <- l$count
l <- estimateMoleculeSize(object)
object@parameters@maxLinkRange <- l$size
object@parameters@breakNCells <- 0.75 * ((object@parameters@maxLinkRange * (object@parameters@maxLinkRange + 1)) / 2)
l <- estimateMinRowCount(object)
object@parameters@minRowCount <- l$count
```

 - Unpolished contigs: /work2/project/seqoccin/assemblies/assemblers/nanopore/bos_taurus/wtdbg2_trio1_mother_37165_run123/mother_raw.cgt.fa
 - ONT: /work2/project/seqoccin/data/reads/nanopore/bos_taurus/trio1.mother.run?.fastq.gz
 - Hi-C: /work2/project/seqoccin/assemblies/scaffolders/hic/bos_taurus/juicer_trio1_mother_37165_wtdbg2_ont_run123_hic_run1/Maison_plus/aligned/inter_30.hic
 - 10X: /work2/project/seqoccin/assemblies/scaffolders/10x/bos_taurus/LR_wtdbg_assembly_mother_37165_run123/Bovin-37165-align/outs/possorted_bam.bam

```{bash}
~/Desktop/Apps/minimap2/minimap2 -t 10 -o /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/GCF_002263795.1_ARS-UCD1.2_genomic.fna.gz /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa
~/Apps/samtools-1.10/samtools faidx /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa
cut -f 1-2 /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa.fai > /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len


~/Apps/samtools-1.10/samtools view -c /Scratch/mazytnicki/Mother_1/possorted_bam.bam
# 1590040440

../CheckScaffolding/checkBreaks /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len 1000000 100000 outBreaks.tsv
../CheckScaffolding/checkBreaks /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len breaks_hic.tsv 100000 10000 predicted.out putative.out > difference_breaks_hic.txt
../CheckScaffolding/checkBreaks /Scratch/mazytnicki/Mother_1/mapping.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.len breaks_ont.tsv 100000 10000 predicted.out putative.out > difference_breaks_ont.txt
../CheckScaffolding/checkChromosome /Scratch/mazytnicki/Mother_1/mapping.paf > scaffChrs.txt
../CheckScaffolding/checkStitch scaffChrs.txt stitch.txt

./checkJoins /Scratch/mazytnicki/Mother_1/mapping.paf joins.txt 10000 10000

zcat /Scratch/mazytnicki/Mother_1/trio1.mother.run1.fastq.gz | wc -l
# 12064000

~/Desktop/Apps/minimap2/minimap2 -t 10 -o /Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa /Scratch/mazytnicki/Mother_1/trio1.mother.run1.fastq.gz

wc -l /Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf
# 4862048 /Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
devtools::load_all(".")
hicFileName <- "/Scratch/mazytnicki/Mother_1/inter_30.hic"
hicData <- parseHicFile(hicFileName, 10000)
hicObject <- tenxcheckerExp(hicData)
hicObject <- estimateDistributions(hicObject)
hicObject@parameters@sampleSize   <- 100000
hicObject@parameters@minCount     <- 3
hicObject@parameters@maxLinkRange <- 500
hicObject@parameters@minRowCount  <- 10
hicObject@parameters@minNBins     <- 2
hicObject@parameters@loessSpan    <- 0.5
hicObject <- cleanData(hicObject)
hicBreaks <- checkBreaks(hicObject)
hicObject@parameters@breakNCells    <- 30
hicObject@parameters@breakThreshold <- -0.3
filteredHicBreaks <- filterBreaks(hicObject, hicBreaks$data)
filteredHicBreaks$breaks %>% mutate(pos = bin * hicObject@parameters@binSize) %>% mutate(ref = str_to_lower(ref)) %>% select(ref, pos) %>% write_tsv("breaks_hic.tsv", col_names = FALSE)
hicObject <- splitChromosomes(hicObject, filteredHicBreaks$breaks)
hicObject@parameters@maxLinkRange <- 10
hicObject@parameters@breakNCells <- 10
joins <- checkJoins(hicObject)
filteredJoins <- filterJoins(hicObject, joins)
hicObject <- scaffold(hicObject, filteredJoins)
plot.10X(hicObject)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
#library("BiocParallel")
#register(MulticoreParam(workers = 10), default = TRUE)
devtools::load_all(".")
bamFileName <- "/Scratch/mazytnicki/Mother_1/possorted_bam.bam"
#bamFileName <- "/Scratch/mazytnicki/verySmallTest.bam"
bamData           <- parseBamFile(bamFileName, 10000)
gc()
bamObject         <- tenxcheckerExp(bamData)
#plotBackgroundCountDistribution(bamObject)
#plotMoleculeSizeDistribution(bamObject)
#plotRowCountDistribution(bamObject)
bamObject         <- estimateDistributions(bamObject)
bamObject@parameters@minCount <- 5
bamObject@parameters@maxLinkRange <- 10
bamObject@parameters@minRowCount <- 100
bamObject@parameters@minNBins <- 2
bamObject@parameters@maxLinkRange   <- 10
bamObject@parameters@breakThreshold <- -3
bamObject@parameters@breakNCells <- 0.75 * ((bamObject@parameters@maxLinkRange * (bamObject@parameters@maxLinkRange + 1)) / 2)
gc()
bamObject         <- cleanData(bamObject)
gc()
bamBreaks         <- checkBreaks(bamObject)
bamObject         <- estimateBreakThreshold(bamObject, bamBreaks$data, 0.05)
filteredBamBreaks <- filterBreaks(bamObject, bamBreaks$data)
bamObject <- splitChromosomes(bamObject, filteredBamBreaks$breaks)
gc()
joins <- checkJoins(bamObject)
gc()
joins <- filterJoins(bamObject, joins)
bamObject <- scaffold(bamObject, joins)
```

```{r}
library("MASS", pos = "package:base")
library("devtools")
library("magrittr")
library("tidyverse")
library("BiocParallel")
library("tibble")
devtools::load_all(".")
ontFileName <- "/Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf"
ontData <- parsePafFile(ontFileName, 10000, 500, 10, 10)
ontObject <- tenxcheckerExp(ontData)
ontObject <- estimateDistributions(ontObject)
ontObject@parameters@minCount <- 2
ontObject@parameters@minRowCount <- 10
ontObject@parameters@minNBins <- 2
plotBackgroundCountDistribution(ontObject)
plotMoleculeSizeDistribution(ontObject)
plotRowCountDistribution(ontObject)
ontObject <- cleanData(ontObject)
ontBreaks <- checkBreaks(ontObject, FALSE, FALSE, TRUE)
ontObject@parameters@maxLinkRange <- 2
ontObject@parameters@breakNCells <- 3
ontObject@parameters@breakThreshold <- -5
ontObject <- estimateBreakThreshold(ontObject, ontBreaks$data, 0.05)
filteredOntBreaks <- filterBreaks(ontObject, ontBreaks$data)
filteredOntBreaks$breaks %>% mutate(pos = bin * ontObject@parameters@binSize) %>% mutate(ref = str_to_lower(ref)) %>% select(ref, pos) %>% write_tsv("breaks_ont.tsv", col_names = FALSE)
ontObject <- splitChromosomes(ontObject, filteredOntBreaks$breaks)
joins <- checkJoins(ontObject)
joins <- filterJoins(ontObject, joins)
ontObject <- scaffold(ontObject, joins)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
devtools::load_all(".")
resolution  <- 10000
hicFileName <- "/Scratch/mazytnicki/Mother_1/inter_30.hic"
hicData     <- parseHicFile(hicFileName, resolution)
gc()
ontFileName <- "/Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf"
ontData     <- parsePafFile(ontFileName, resolution, 500, 10, 10)
gc()
bamFileName <- "/Scratch/mazytnicki/Mother_1/possorted_bam.bam"
bamData     <- parseBamFile(bamFileName, resolution)
gc()
allObject <- tenxchecker(resolution)
allObject <- addExp(allObject, ontData, "ONT")
allObject <- addExp(allObject, hicData, "HiC")
allObject <- addExp(allObject, bamData, "10X")
gc()
allObject <- estimateDistributions(allObject)
gc()
allObject <- cleanData(allObject)
gc()
allObject <- findBreaks(allObject, 0.05)
allObject <- splitChromosomes(allObject)
allObject <- findJoins(allObject, 0.05)
allObject <- scaffold(allObject)
allObject@joins %>% select(-pvalue) %>% write_tsv("joins_100k.tsv", col_names = FALSE)
```

```{bash}
~/Apps/samtools-1.10/samtools index -@ 5 /Scratch/mazytnicki/Mother_1/possorted_bam.bam
~/Apps/samtools-1.10/samtools view -o /Scratch/mazytnicki/Mother_1/possorted_bam_reduced.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg10 ctg1061 ctg118 ctg1185 ctg12 ctg126 ctg128 ctg13 ctg1332 ctg136 ctg1427 ctg1435 ctg1451 ctg146 ctg154 ctg17 ctg173 ctg18 ctg189 ctg198 ctg2112 ctg219 ctg22 ctg23 ctg2379 ctg257 ctg26 ctg276 ctg281 ctg29 ctg30 ctg31 ctg312 ctg32 ctg322 ctg33 ctg3563 ctg3623 ctg38 ctg4 ctg4032 ctg4123 ctg4149 ctg43 ctg4330 ctg446 ctg4519 ctg4700 ctg4854 ctg4913 ctg50 ctg5008 ctg514 ctg516 ctg531 ctg549 ctg551 ctg5765 ctg6 ctg61 ctg655 ctg669 ctg67 ctg68 ctg703 ctg737 ctg766 ctg79 ctg8 ctg80 ctg82 ctg9 ctg91 ctg95 ctg976
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg4_ctg17.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg4 ctg17
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg2981_ctg1688.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg1688 ctg2981
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/small_test.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg104 ctg225
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg11.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg11
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg100.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg100
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg90.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg90
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg188.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg188
~/Apps/samtools-1.10/samtools view -@ 6 -o /Scratch/mazytnicki/Mother_1/possorted_bam_ctg37.bam /Scratch/mazytnicki/Mother_1/possorted_bam.bam ctg37
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("Biostrings")
devtools::load_all(".")
resolution  <- 1000
 hicFileName <- "/Scratch/mazytnicki/Mother_1/inter_30.hic"
 hicData     <- parseHicFile(hicFileName, resolution)
 ontFileName <- "/Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf"
 ontData     <- parsePafFile(ontFileName, resolution, 500, 10, 10)
bamFileName <- "/Scratch/mazytnicki/Mother_1/possorted_bam_reduced.bam"
bamData     <- parseBamFile(bamFileName, resolution)
contigFileName <- "/Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa"
allObject <- tenxchecker(contigFileName, resolution)
 allObject <- addExp(allObject, ontData, "ONT")
 allObject <- addExp(allObject, hicData, "HiC")
allObject <- addExp(allObject, bamData, "10X")
allObject <- keepScaffolds(allObject, c("ctg10", "ctg1061", "ctg118", "ctg1185", "ctg12", "ctg126", "ctg128", "ctg13", "ctg1332", "ctg136", "ctg1427", "ctg1435", "ctg1451", "ctg146", "ctg154", "ctg17", "ctg173", "ctg18", "ctg189", "ctg198", "ctg2112", "ctg219", "ctg22", "ctg23", "ctg2379", "ctg257", "ctg26", "ctg276", "ctg281", "ctg29", "ctg30", "ctg31", "ctg312", "ctg32", "ctg322", "ctg33", "ctg3563", "ctg3623", "ctg38", "ctg4", "ctg4032", "ctg4123", "ctg4149", "ctg43", "ctg4330", "ctg446", "ctg4519", "ctg4700", "ctg4854", "ctg4913", "ctg50", "ctg5008", "ctg514", "ctg516", "ctg531", "ctg549", "ctg551", "ctg5765", "ctg6", "ctg61", "ctg655", "ctg669", "ctg67", "ctg68", "ctg703", "ctg737", "ctg766", "ctg79", "ctg8", "ctg80", "ctg82", "ctg9", "ctg91", "ctg95", "ctg976"))
allObject <- estimateDistributions(allObject)
allObject@data[[3]]@parameters@maxLinkRange <- 50
allObject <- cleanData(allObject)
 allObject <- findBreaks(allObject, 0.05)
 allObject <- splitChromosomes(allObject)
allObject <- findJoins(allObject, 0.00001)
allObject <- scaffold(allObject)
writeXStringSet(allObject@sequences, "sequences.fa")


allObject@data[[1]]@joins@data
ggsave("tmp.png", allObject@data[[1]]@joins@testPlots[[1]])
ggsave("tmp2.png", allObject@data[[1]]@joins@mapPlots[[1]])
allObject@sizes$ctg30
allObject@sizes$ctg127
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("ggforce")
devtools::load_all(".")
resolution  <- 100
bamFileName <- "/Scratch/mazytnicki/Mother_1/small_test.bam"
bamData     <- parseBamFile(bamFileName, resolution)
allObject <- tenxchecker(resolution)
allObject <- addExp(allObject, bamData, "10X")
allObject <- estimateDistributions(allObject)
allObject <- cleanData(allObject)
ggsave("tmp0.png", plot.10X(allObject@data[[1]], allObject@sizes))
allObject@data[[1]]@parameters@maxLinkRange <- 100
allObject <- findJoins(allObject, 0.05)
ggsave("tmp.png", allObject@data[[1]]@joins@testPlots[[1]])
ggsave("tmp2.png", allObject@data[[1]]@joins@mapPlots[[1]])
allObject <- scaffold(allObject)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("Biostrings")
devtools::load_all(".")
resolution  <- 100
bamFileName <- "/Scratch/mazytnicki/Mother_1/small_test.bam"
contigFileName <- "ctg104_ctg225.fa"
bamData     <- parseBamFile(bamFileName, resolution)
allObject <- tenxchecker(contigFileName, resolution)
allObject <- addExp(allObject, bamData, "10X")
allObject <- estimateDistributions(allObject)
allObject@data[[1]]@parameters@maxLinkRange <- 50
allObject <- cleanData(allObject)
allObject <- findBreaks(allObject, 0.05)
allObject <- findJoins(allObject, 0.05)
allObject <- scaffold(allObject)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("Biostrings")
devtools::load_all(".")
resolution  <- 100
bamFileName <- "/Scratch/mazytnicki/Mother_1/possorted_bam_ctg37.bam"
bamData     <- parseBamFile(bamFileName, resolution)
contigFileName <- "/Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa"
allObject <- tenxchecker(contigFileName, resolution)
allObject <- addExp(allObject, bamData, "10X")
allObject <- keepScaffolds(allObject, c("ctg37"))
allObject <- estimateDistributions(allObject)
allObject@data[[1]]@parameters@maxLinkRange <- 50
allObject <- cleanData(allObject)
allObject <- findBreaks(allObject, 0.05)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("Biostrings")
devtools::load_all(".")
resolution  <- 1000
contigFileName <- "/Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa"
hicFileName <- "/Scratch/mazytnicki/Mother_1/inter_30.hic"
hicData     <- parseHicFile(hicFileName, resolution)
allObject <- tenxchecker(contigFileName, resolution)
allObject <- addExp(allObject, hicData, "HiC")
allObject <- keepScaffolds(allObject, c("ctg10", "ctg1061", "ctg118", "ctg1185", "ctg12", "ctg126", "ctg128", "ctg13", "ctg1332", "ctg136", "ctg1427", "ctg1435", "ctg1451", "ctg146", "ctg154", "ctg17", "ctg173", "ctg18", "ctg189", "ctg198", "ctg2112", "ctg219", "ctg22", "ctg23", "ctg2379", "ctg257", "ctg26", "ctg276", "ctg281", "ctg29", "ctg30", "ctg31", "ctg312", "ctg32", "ctg322", "ctg33", "ctg3563", "ctg3623", "ctg38", "ctg4", "ctg4032", "ctg4123", "ctg4149", "ctg43", "ctg4330", "ctg446", "ctg4519", "ctg4700", "ctg4854", "ctg4913", "ctg50", "ctg5008", "ctg514", "ctg516", "ctg531", "ctg549", "ctg551", "ctg5765", "ctg6", "ctg61", "ctg655", "ctg669", "ctg67", "ctg68", "ctg703", "ctg737", "ctg766", "ctg79", "ctg8", "ctg80", "ctg82", "ctg9", "ctg91", "ctg95", "ctg976"))
allObject <- estimateDistributions(allObject)
allObject <- cleanData(allObject)
#allObject <- findBreaks(allObject)
#allObject <- splitChromosomes(allObject)
allObject <- findJoins(allObject, 0.05)
allObject@data[[1]]@joins@data
ggsave("tmp.png", allObject@data[[1]]@joins@testPlots[[1]])
ggsave("tmp2.png", allObject@data[[1]]@joins@mapPlots[[1]])
allObject@sizes$ctg30
allObject@sizes$ctg127
allObject <- findJoins(allObject, 0.05)
allObject@joins %>% select(ref1, ref2) %>% write_tsv("joins.tsv", col_names = FALSE)
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("Biostrings")
library("cowplot")
devtools::load_all(".")
resolution  <- 1000
bamFileName <- "/Scratch/mazytnicki/Mother_1/possorted_bam2.bam"
bamData     <- parseBamFile(bamFileName, resolution)
hicFileName <- "/Scratch/mazytnicki/Mother_1/inter_30.hic"
hicData     <- parseHicFile(hicFileName, resolution)
ontFileName <- "/Scratch/mazytnicki/Mother_1/trio1.mother.run1.paf"
ontData     <- parsePafFile(ontFileName, resolution, 500, 10, 10)
contigFileName <- "/Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa"
allObject <- tenxchecker(contigFileName, resolution)
allObject <- addExp(allObject, ontData, "ONT")
allObject <- addExp(allObject, hicData, "HiC")
allObject <- addExp(allObject, bamData, "10X")
allObject <- estimateDistributions(allObject)
allObject@data[[3]]@parameters@maxLinkRange <- 50
allObject <- cleanData(allObject)
#saveRDS(allObject, file = "/Scratch/mazytnicki/Mother_1/allObject.rds")
#allObject <- readRDS(file = "/Scratch/mazytnicki/Mother_1/allObject.rds")
allObject <- findBreaks(allObject, 0.05)
gc()
allObject <- splitChromosomes(allObject)
gc()
allObject <- findJoins(allObject, 0.05)
allObject <- scaffold(allObject)
writeXStringSet(allObject@sequences, "sequences.fa")
allObject@joins %>% select(ref1, ref2) %>% write_tsv("joins.tsv", col_names = FALSE)
```

```{bash}
../CheckScaffolding/verifyJoins /Scratch/mazytnicki/Mother_1/mapping.paf 100000 10000 joins.tsv

# From tapou
~/Desktop/Apps/quast-5.0.2/quast-lg.py -t 15 -r /Scratch/mazytnicki/GCF_002263795.1_ARS-UCD1.2_genomic.fna.gz -t 6 -o /home/mazytnicki/Desktop/Projects/Other/Assembly/Quast/Test1 /Scratch/mazytnicki/Mother_1/mother_raw.cgt.fa wtdbg2_trio1_mother_37165_arcs_links.fa sequences.fa sequences_ONT_HiC.fa
# From genologin
```

```{r}
library("devtools")
library("magrittr")
library("tidyverse")
library("Biostrings")
library("cowplot")
devtools::load_all(".")
#object <- makeSimObject(1, 1, 100, 20, 5, 0.2, c(1), c(50), c(), c(), c())
object <- makeSimObject(2, 5, 100, 20, 10, 0.2, c(1, 2), c(25, 75), c(3, 4), c(4, 5), c(1, 5))
object <- makeSimObject(2, 5, 100, 20, 10, 0.2, c(1, 1, 2, 2), c(25, 75, 75, 75), c(3, 4), c(4, 5), c(1, 5))
object <- makeSimObject(1, 2, 100, 20, 10, 0.2, c(), c(), c(1), c(2), c(5))
plot.10X(object)
object <- estimateDistributions(object)
object@data[[1]]@parameters@maxLinkRange <- 6
object@data[[2]]@parameters@maxLinkRange <- 6
object <- cleanData(object)
object <- findBreaks(object, 0.05)
object <- splitChromosomes(object)
object <- findJoins(object, 0.05)
```

Erreur dans extractLines(object@interactionMatrix, lines, object@parameters@maxLinkRange) :
Error in 'extractLines', got negative distances: ref1: 1, ref2: 1, bin1: 1, bin2: 365.


object@data[[1]]@interactionMatrix %>% filter(ref1 == "ref_2", ref2 == "ref_1") %>% dplyr::mutate(bin2 = 99 - bin2) -> interactionMatrix
interactionMatrix %>% filter(bin1 <= 10, bin2 <= 10) %>% ggplot(aes(x = bin1, y = bin2, fill = count)) + geom_raster()
interactionMatrix %>% ggplot(aes(x = bin1, y = bin2, fill = count)) + geom_raster()

maxDistance <- object@data[[1]]@parameters@maxLinkRange
isCorner <- TRUE
bins <- seq.int(from = 0, to = maxDistance, by = 1)                                                                                                                                                            
d    <- function(bin1, bin2, isCorner) {
	if (isCorner) {
		return(bin1 + bin2)
	}
		return(bin1 - bin2)
}
emptyMatrix <- tibble(bin1 = bins, bin2 = bins) %>%
	tidyr::expand(bin1, bin2) %>%
	dplyr::mutate(distance = d(bin1, bin2, isCorner)) %>%
	dplyr::filter(distance >= 0) %>%
	dplyr::filter(distance <= maxDistance)                                                                                                                                                                     
emptyMatrix %>%
	dplyr::left_join(interactionMatrix, by = c("bin1", "bin2")) %>%
	tidyr::replace_na(list(count = 0)) %>%
        dplyr::mutate(count = as.integer(count)) -> interactionMatrix2


interactionMatrix2 %>% ggplot(aes(x = bin1, y = bin2, fill = count)) + geom_raster()

offset <- 0
f <- function(offset) {
	object@data[[1]]@parameters@distanceCount %>% mutate(distance = distance - offset) %>% filter(distance >= 0) -> background
	interactionMatrix2 %>% filter(distance <= maxDistance - offset) %>% dplyr::left_join(background, by = "distance") %>% dplyr::mutate(d = (count - meanCount))
}
map_dfr(0:13, f, .id = "offset") %>% mutate(offset = as.integer(offset)) %>% ggplot(aes(x = bin1, y = bin2, fill = d)) + geom_raster() + scale_fill_gradient2() + facet_grid(cols = vars(offset))

g <- function(offset) {
	object@data[[1]]@parameters@distanceCount %>% mutate(distance = distance - offset) %>% filter(distance >= 0) -> background
	interactionMatrix2 %>% filter(distance <= maxDistance - offset) %>% dplyr::left_join(background, by = "distance") %>% dplyr::mutate(d = abs(count - meanCount)) %>% pull(d) %>% median() -> r
	#interactionMatrix2 %>% filter(distance <= maxDistance - offset) %>% dplyr::left_join(background, by = "distance") %>% dplyr::mutate(d = (count - meanCount) ^ 2) %>% pull(d) %>% sum() %>% sqrt() -> r
        return(r / nrow(interactionMatrix2))
}
map_dbl(0:13, g)

emptyMatrix <- tibble(bin1 = bins, bin2 = bins) %>%
	tidyr::expand(bin1, bin2) %>%
	dplyr::mutate(distance = bin1 - bin2) %>%
	dplyr::filter(distance >= 0) %>%
	dplyr::filter(distance <= maxDistance)                                                                                                                                                                     
object@data[[1]]@interactionMatrix %>%
	filter(ref1 == ref2) %>%
	mutate(distance = bin1 - bin2) %>%
        right_join(emptyMatrix, by = c("bin1", "bin2", "distance")) %>%
	tidyr::replace_na(list(count = 0)) %>%
	select(distance, count) -> fullInteractionMatrix
lo <- loess(count ~ distance, data=fullInteractionMatrix)
fullInteractionMatrix$loess <- predict(lo)

ggplot(fullInteractionMatrix, aes(x = distance, y = count)) + geom_bin2d(binwidth = 1) + geom_point(aes(y = loess), col="red")

lo <- loess(count ~ distance, data=interactionMatrix2)
interactionMatrix2$loess <- predict(lo)

ggplot(interactionMatrix2, aes(x = distance, y = count)) + geom_bin2d(binwidth = 1) + geom_point(aes(y = loess), col="red")

bLoess <- fullInteractionMatrix %>% select(distance, loess) %>% arrange(distance) %>% distinct()
oLoess <- interactionMatrix2 %>% select(distance, loess) %>% arrange(distance) %>% distinct()

h <- function(offset) {
	bLoess %>% mutate(distance = distance - offset) %>% filter(distance >= 0) -> background
	oLoess %>% filter(distance <= maxDistance - offset) %>% dplyr::left_join(background, by = "distance", suffix = c("_o", "_b")) %>% dplyr::mutate(d = abs(loess_b - loess_o)) %>% pull(d) %>% sum() -> r
        return(r / (maxDistance - offset + 1))
}
map_dbl(0:13, h)
